sdk/include/opentelemetry/sdk/trace/span_data.h

```diff
@@ -106,6 +121,7 @@ class SpanData final : public Recordable
   std::string name_;
   opentelemetry::trace::CanonicalCode status_code_{opentelemetry::trace::CanonicalCode::OK};
   std::string status_desc_;
+  std::unordered_map<std::string, common::AttributeValue> attributes_;
```

Comment:
Why does this have to be a string? We can add order comparators to nostd::string_view implementation, then you do not need to construct std::string object to transform the nostd::string_view key into std::string... Standard Library has comparators on std::string_view , allowing it to be used as key in the maps.

Reply:
This is more of a lifetime issue. `string_view` is a non-owning reference, so the value pointed to by a `string_view` might be invalidated (freed) while we hold on to it.

Reply:
We do have the same issue with the rest of API surface.

For example, _AddEvent_ on span, we we use **string_view** for keys.

My concern as follows:
- for flows that require synchronous exporter, such as Windows **ETW**, or anything that involves "local" high-speed RPC channel, e.g. **fluentd** - we do not necessarily need to copy the key.

- also for async flows: if the key is a string view of a constant character literal, such as (pseudocode):
```
SetAttribute("MyKey1", ...);
SetAttribute("MyKey2", ...);
```

or

```
    M m2 = {{"key1", "one"}, {"key2", "two"}};
    span->AddEvent("MyProduct.MyEvent2", m2);
```

string literal keys would permanently live in read-only **.data** , and copying them is wasteful.

Should this be spelled out in some documentation, when / why the memcpy is necessary for the keys?

Reply:
> * for flows that require synchronous exporter, such as Windows ETW, or anything that involves "local" high-speed RPC channel, e.g. fluentd - we do not necessarily need to copy the key.

For those cases implementing a custom `Recordable` makes sense. The exporter implementor has full control over what data is copied, the SDK makes no copies whatsoever.

> string literal keys would permanently live in read-only .data , and copying them is wasteful.

True. But on exporter side, it would be hard to determine which of the strings received live in read-only **.data** and which are allocated on the heap/stack. It also would be complicated to come up with an optimization that accounts for that.

I'm hesitant to make `SpanData` a highly optimized part of the SDK, as vendors with high performance requirements have the possibility to come up with `Recordable`s tailored to their needs. With the `Recordable` approach, we don't need to come up with one `SpanData` that fits all needs (which is impossible anyway).